src/System.Management.Automation/engine/Attributes.cs

```diff
@@ -1455,6 +1533,44 @@ public ValidateSetAttribute(params string[] validValues)
 
             _validValues = validValues;
         }
+
+        /// <summary>
+        /// Initializes a new instance of the ValidateSetAttribute class.
+        /// Valid values is returned dynamically from a custom class implementing 'IValidateSetValuesGenerator' interface.
+        /// </summary>
+        /// <param name="valuesGeneratorType">class that implements the 'IValidateSetValuesGenerator' interface</param>
+        /// <exception cref="ArgumentException">for null arguments</exception>
+        public ValidateSetAttribute(Type valuesGeneratorType)
+        {
+            if (!typeof(IValidateSetValuesGenerator).IsAssignableFrom(valuesGeneratorType))
```

Comment:
I think basic validations are necessary. For example, do you allow people to use an internal/private type as an argument? `Activator.CreateInstance` works on non-public types but I don't think we should support non-public types.